README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

[![Travis build
status](https://travis-ci.org/CorradoLanera/weektwo.svg?branch=master)](https://travis-ci.org/CorradoLanera/weektwo)
[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/CorradoLanera/weektwo?branch=master&svg=true)](https://ci.appveyor.com/project/CorradoLanera/weektwo)
[![Coverage
status](https://codecov.io/gh/CorradoLanera/weektwo/branch/master/graph/badge.svg)](https://codecov.io/github/CorradoLanera/weektwo?branch=master)

[![lifecycle](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![CRAN
status](https://www.r-pkg.org/badges/version/weektwo)](https://cran.r-project.org/package=weektwo)

# weektwo

The goal of **weektwo** is to demonstrate the ability to create a
package.

## Installation

You can install the develop version of **weektwo** from
[GITHUB](https://github.com) with:

``` r
# install.packages("devtools")
devtools::install_github("CorradoLanera/weektwo")
```

## Example

``` r
library(weektwo)

path_to_data <- system.file("sample-data", package = "weektwo")

fars_map_state(state.num = 1,
    year = 2013,
    path = path_to_data
)
```

<img src="man/figures/README-unnamed-chunk-1-1.png" width="100%" />

## Code of Conduct

Please note that the **weektwo** project is released with a [Contributor
Code of Conduct](CODE_OF_CONDUCT.md). By contributing to this project,
you agree to abide by its terms.


NEWS.md
# weektwo 0.1

* assignment compelete, build pass with 0 errors, 0 warnings, 0 notes

# weektwo 0.0.3

* update `README`
* create vignette `weektwo-basics`

# weektwo 0.0.2

* convert `fars_summarize_years()` to SE.

# weektwo 0.0.1

* move sample data in the `test` folder
* introduce the use of `glue::glue()`
* introduce some `assertive` checks
* write tests for all the functions
* all function in its own files

# weektwo 0.0.0.9000

* Add `inst/sample-data` and move sample data into it
* Add a sample vignette
* Add basic test for pipe
* Add standard support (CoC, LICENSE, README, NEWS, testthat, pipe, 
  travis, appveyor, codecov, CONTRIBUTING, ISSUE_TEMPLATE, SUPPORT)


tests/testthat.R
library(testthat)
library(weektwo)

test_check("weektwo")


tests/testthat/test-fars_read.R
context("test-fars_read")

readed_data <- fars_read("accident_2013.csv.bz2")

test_that("import works", {

    expect_is(readed_data, "data.frame")

    expect_equal(
        readed_data[["ST_CASE"]][[1]],
        10001L
    )

})


tests/testthat/test-pipe-works.R
context("test-pipe-works")

test_that("simple pipe", {
  expect_equal(1 %>% sum(2), 3)
})


R/fars_read.R
#' Read fars files
#'
#' Read a fars file returning the imported
#' \code{\link[dplyr]{data_frame}}.
#'
#' An error is returned if the path is not valid.
#'
#' @details Fars file has to be stored in a \code{CSV} file format.
#'   It can be possibly compressed (see \code{\link{file}}).
#'
#' @param filename (chr) Complete path to the fars file to read.
#'
#' @return A \code{\link[dplyr]{data_frame}}.
#' @export
#'
#' @examples
#' \dontrun{
#'    file_path <- system.file("tests", "testthat", package = "weektwo",
#'        "accident_2013.csv.bz2"
#'    )
#'    fars_read(file_path)
#' }
fars_read <- function(filename) {

    if (!file.exists(filename)) {
        stop("file '", filename, "' does not exist in", getwd())
    }

    suppressMessages({
        readr::read_csv(filename, progress = FALSE)
    })
}


R/fars_functions.R
#' Read fars
#'
#' Read a fars file  returning the imported \code{data_frame}. An error is
#' returned if the path is not valid.
#'
#' @details Fars file has to be stored in a \code{CSV} file format.
#'   It can be possibly compressed (see \code{\link{file}}).
#'
#' @param filename (chr) Complete path to the fars file to read.
#'
#' @importFrom readr read_csv
#' @importFrom dplyr tbl_df
#'
#' @return A \code{data_frame}.
#' @export
#'
#' @examples
#' \dontrun{
#'     library(weektwo)
#'     sample_data <- system.file("sample-data/accident_2013.csv.bz2",
#'         package = "weektwo"
#'     )
#'
#'     fars_read(sample_data)
#' }
fars_read <- function(filename) {
    if (!file.exists(filename))
        stop("file '", filename, "' does not exist")
    data <- suppressMessages({
        readr::read_csv(filename, progress = FALSE)
    })
    dplyr::tbl_df(data)
}

#' Make fars filename
#'
#' Compose the correct name for a fars file given the reference year.
#'
#' @note This function is for internal usage only and it is not
#'       exported.
#'
#' @param year (num) The year of the interested fars file.
#'
#' @return A character string of length one with the filename
#'     (full extension included) of the fars file for the year of
#'     interest.
#'
#' @examples
#' \dontrun{
#'     weektwo:::make_filename("2012")
#' }
make_filename <- function(year) {
    year <- as.integer(year)
    sprintf("accident_%d.csv.bz2", year)
}


#' Read multiple fars
#'
#' Import a set of fars files given the vector of their years.
#'
#' @details Computation does not stop if the file does not exist: a
#'    warning is produced and \code{NULL} is stored in the corresponding
#'    slot of the output list.
#'
#' @param years (num) An integer vector of years (or an object
#'        cohercible to it) containing the years of fars files
#'        of interest.
#'
#' @importFrom dplyr mutate select
#'
#' @return A list of \code{data_frame}, one for each year privided in
#'         input, with two columns: \code{MONTH}, and \code{year} of the
#'         corresponding fars file. If the file does not exist
#'         \code{NULL} is stored in the corresponding slot of the list.
#' @export
#'
#' @examples
#' \dontrun{
#'     fars_read_years(c(2012, 2013, 2014))                # one warning
#'     fars_read_years(c(2013, 2014))                       # no warning
#'     fars_read_years(c("2012", "2013", "2014"))# works with characters
#' }
fars_read_years <- function(years) {
    lapply(years, function(year) {
        file <- make_filename(year)
        tryCatch({
            dat <- fars_read(file)
            dplyr::mutate(dat, year = year)[, c("MONTH", "year")]
        }, error = function(e) {
            warning("invalid year: ", year)
            return(NULL)
        })
    })
}

#' Summarize fars record by year
#'
#' Compute the number of records are present for fars for each year
#' (columns) and month (rows).
#'
#' @note If a year not present in the data is provided, a warning is
#'       returned and no column is created for it.
#'
#' @inheritParams fars_read_years
#'
#' @return A \code{data_frame} with a number of column equals to the
#'         number of (valid) years passed as input (plus one for the
#'         months) and one row each month included in the corresponding
#'         fars files. The \code{data_frame} is filled with the number
#'         of records included for the corresponding year and month.
#'
#' @importFrom dplyr bind_rows group_by summarize
#' @importFrom tidyr spread
#'
#' @export
#'
#' @examples
#' \dontrun{
#'     fars_summarize_years(c(2013, 2014))
#'     fars_summarize_years(c(2012, 2013, 2014))
#' }
fars_summarize_years <- function(years) {
    dat_list <- fars_read_years(years)
    dplyr::bind_rows(dat_list) %>%
        dplyr::group_by(year, MONTH) %>%
        dplyr::summarize(n = n()) %>%
        tidyr::spread(year, n)
}

#' Map accidents
#'
#' Plot the point map for the fars record of the STATE corresponding to
#' the \code{state.num} and the \code{year} provided
#'
#' @details if no accident happen in the state and period selected
#'          a message will be throw and \code{NULL} is returned.
#'
#'          If a plot is returned, it will be drawn only for longitude
#'          not greater than 900 and latidude not grater than 90.
#'
#' @inheritParams make_filename
#' @param state.num (num) An integer (or something cohercible to it)
#'        representing a STATE number into the fars data.
#'
#' @return \code{NULL} (with a message) if the STATE represented by the
#'         \code{state.num} provided has no records in fars for the year
#'         selected. Otherwhise, a plot with the geographical point for
#'         its fars is produced and nothing will be returned.
#'
#'
#' @importFrom dplyr filter
#' @importFrom maps map
#' @importFrom graphics points
#'
#' @export
#'
#' @examples
#' \dontrun{
#'     fars_map_state(1, 2013)
#'     fars_map_state(1, 2012) # error
#'     fars_map_state(2, 2015) # nothing to drawn
#' }
fars_map_state <- function(state.num, year) {
    filename <- make_filename(year)
    data <- fars_read(filename)
    state.num <- as.integer(state.num)

    if (!(state.num %in% unique(data[["STATE"]])))
        stop("invalid STATE number: ", state.num)
    data.sub <- dplyr::filter(data, STATE == state.num)
    if (nrow(data.sub) == 0L) {
        message("no accidents to plot")
        return(invisible(NULL))
    }
    is.na(data.sub$LONGITUD) <- data.sub$LONGITUD > 900
    is.na(data.sub$LATITUDE) <- data.sub$LATITUDE > 90
    with(data.sub, {
        maps::map("state", ylim = range(LATITUDE, na.rm = TRUE),
            xlim = range(LONGITUD, na.rm = TRUE))
        graphics::points(LONGITUD, LATITUDE, pch = 46)
    })
}


R/fars_read_years.R
#' Read multiple fars
#'
#' Import a set of fars files given the vector of their years.
#'
#' @details Computation does not stop if the file does not exist: a
#'    warning is produced and \code{NULL} is stored in the corresponding
#'    slot of the output list.
#'
#' @param years (num) An integer vector of years (or an object
#'        cohercible to it) containing the years of fars files
#'        of interest.
#' @param path (chr) path to the fars file
#'
#' @return A list of data frame, one for each year privided in
#'         input, with two columns: \code{MONTH}, and \code{YEAR} of the
#'         corresponding fars file. If the file does not exist
#'         \code{NULL} is stored in the corresponding slot of the list.
#' @export
#'
#' @examples
#' \dontrun{
#'     file_path <- system.file("tests", "testthat",
#'         package = "weektwo"
#'     )
#'
#'     fars_read_years(c(2012, 2013, 2014), path = file_path) # one warn
#'     fars_read_years(c(2013, 2014)) # no warning
#'     fars_read_years(c("2012", "2013", "2014")) # works with characters
#' }
fars_read_years <- function(years, path = ".") {

    years <- suppressWarnings(
        purrr::set_names(as.numeric(years), as.numeric(years))
    )

    assertive::assert_all_are_whole_numbers(years)

    purrr::map(years, function(year) {

        file_path <- file.path(path, make_filename(year))

        tryCatch(
            fars_read(file_path)[c("MONTH", "YEAR")],
            error = function(e) {
                warning("invalid year: ", year, call. = FALSE)
                return(NULL)
            }
        )
    })

}


R/fars_summarize_years.R
#' Summarize fars record by year
#'
#' Compute the number of records are present for fars for each year
#' (columns) and month (rows).
#'
#' @note If a year not present in the data is provided, a warning is
#'       returned and no column is created for it.
#'
#' @inheritParams fars_read_years
#'
#' @return A data frame with a number of column equals to the
#'         number of (valid) years passed as input (plus one for the
#'         months) and one row each month included in the corresponding
#'         fars files. The data frame is filled with the number
#'         of records included for the corresponding year and month.
#'
#' @export
#'
#' @examples
#' \dontrun{
#'    file_path <- system.file("tests", "testthat", package = "weektwo")
#'
#'    fars_summarize_years(c(2013, 2014), path = file_path)
#'
#'    fars_summarize_years(c(2012, 2013, 2014), path = file_path)
#' }
fars_summarize_years <- function(years, path = ".") {

    years <- suppressWarnings(as.numeric(years))
    assertive::assert_all_are_whole_numbers(years)

    dat_list <- fars_read_years(years, path = path)

    dplyr::bind_rows(dat_list) %>%
        dplyr::group_by_("YEAR", "MONTH") %>%
        dplyr::summarize_("n" = "n()") %>%
        tidyr::spread_("YEAR", "n")
}


R/fars_map_state.R
#' Map accidents
#'
#' Plot the point map for the fars record of the STATE corresponding to
#' the \code{state.num} and the \code{year} provided
#'
#' @details if no accident happen in the state and period selected
#'          a message will be throw and \code{NULL} is returned.
#'
#'          If a plot is returned, it will be drawn only for longitude
#'          not greater than 900 and latidude not grater than 90.
#'
#' @inheritParams make_filename
#' @param state.num (num) An integer (or something cohercible to it)
#'        representing a STATE number into the fars data.
#' @param path (chr) path to the fars file
#'
#' @return \code{NULL} (with a message) if the STATE represented by the
#'         \code{state.num} provided has no records in fars for the year
#'         selected. Otherwhise, a plot with the geographical point for
#'         its fars is produced and nothing will be returned.
#'
#'
#' @export
#'
#' @examples
#' \dontrun{
#'     file_path <- system.file("tests", "testthat",
#'         package = "weektwo"
#'     )
#'
#'     fars_map_state(1, 2013, path = file_path)
#'     fars_map_state(1, 2012, path = file_path) # error
#'     fars_map_state(2, 2015, path = file_path) # nothing to drawn
#' }
fars_map_state <- function(state.num, year, path = ".") {

    file_path <- file.path(path, make_filename(year))
    data <- fars_read(file_path)
    state.num <- as.integer(state.num)

    if (!(state.num %in% unique(data[["STATE"]])))
        stop("invalid STATE number: ", state.num)
    data.sub <- data[data[["STATE"]] == state.num, , drop = FALSE]
    if (nrow(data.sub) == 0L) {
        message("no accidents to plot")
        return(invisible(NULL))
    }
    is.na(data.sub$LONGITUD) <- data.sub$LONGITUD > 900
    is.na(data.sub$LATITUDE) <- data.sub$LATITUDE > 90
    with(data.sub, {
        maps::map("state", ylim = range(LATITUDE, na.rm = TRUE),
                  xlim = range(LONGITUD, na.rm = TRUE))
        graphics::points(LONGITUD, LATITUDE, pch = 46)
    })
}
